csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/CXX_Api_Sample.cpp

```diff
@@ -0,0 +1,126 @@
+    input_node_names[i] = input_name;
+
+    // print input node types
+    Ort::TypeInfo type_info=session.GetInputTypeInfo(i);
+    auto tensor_info=type_info.GetTensorTypeAndShapeInfo();
+
+    tensor_info.NonConstTest();
+
+    ONNXTensorElementDataType type = tensor_info.GetElementType();
+    printf("Input %d : type=%d\n", i, type);
+
+    // print input shapes/dims
+    input_node_dims=tensor_info.GetShape();
+    printf("Input %d : num_dims=%zu\n", i, input_node_dims.size());
+    for (int j = 0; j < input_node_dims.size(); j++)
+      printf("Input %d : dim %d=%jd\n", i, j, input_node_dims[j]);
+  }
+
+  // Results should be...
+  // Number of inputs = 1
+  // Input 0 : name = data_0
+  // Input 0 : type = 1
+  // Input 0 : num_dims = 4
+  // Input 0 : dim 0 = 1
+  // Input 0 : dim 1 = 3
+  // Input 0 : dim 2 = 224
+  // Input 0 : dim 3 = 224
+
+  //*************************************************************************
+  // Similar operations to get output node information.
+  // Use OrtSessionGetOutputCount(), OrtSessionGetOutputName()
+  // OrtSessionGetOutputTypeInfo() as shown above.
+
+  //*************************************************************************
+  // Score the model using sample data, and inspect values
+
+  size_t input_tensor_size = 224 * 224 * 3;  // simplify ... using known dim values to calculate size
+                                             // use OrtGetTensorShapeElementCount() to get official size!
+
+  std::vector<float> input_tensor_values(input_tensor_size);
+  std::vector<const char*> output_node_names = {"softmaxout_1"};
+
+  // initialize input data with values in [0.0, 1.0]
+  for (unsigned int i = 0; i < input_tensor_size; i++)
+    input_tensor_values[i] = (float)i / (input_tensor_size + 1);
+
+  // create input tensor object from data values
+  Ort::AllocatorInfo allocator_info = Ort::AllocatorInfo::Create_Cpu(OrtArenaAllocator, OrtMemTypeDefault);
+  Ort::Value input_tensors[1]={Ort::Value::CreateTensor(allocator_info, input_tensor_values.data(), input_tensor_size * sizeof(float), input_node_dims.data(), 4, ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT)};
+  assert(input_tensors[0].IsTensor());
+
+  // score model & input tensor, get back output tensor
+  Ort::Value output_tensor=session.Run(nullptr, input_node_names.data(), input_tensors, output_node_names.data(), 1);
+  assert(output_tensor.IsTensor());
+
+  // Get pointer to output tensor float values
+  float* floatarr=output_tensor.GetTensorMutableData<float>();
+  assert(abs(floatarr[0] - 0.000045) < 1e-6);
+
+  // score the model, and print scores for first 5 classes
+  for (int i = 0; i < 5; i++)
+    printf("Score for class [%d] =  %f\n", i, floatarr[i]);
+
+  // Results should be as below...
+  // Score for class[0] = 0.000045
+  // Score for class[1] = 0.003846
+  // Score for class[2] = 0.000125
+  // Score for class[3] = 0.001180
+  // Score for class[4] = 0.001317
+  printf("Done!\n");
+  return 0;
```

Comment:
This sample would be the first thing new users try out. Should we add a unit test for it, similar to the C-API example? That way it can be run as part of CI if needed.  The runtest.bat command is used for C-API -- but can be reused for C++ as well.

https://github.com/Microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/runtest.bat

Reply:
I was hoping to talk to you about that. How does 'runtest' work? I didn't see C_Api_Sample.cpp anywhere in it.

Reply:
There are a few main steps (see below). The first argument on the command line is the folder containing the NuGet package generated from the build. This test primarily checks  the Native C++ dlls, the C_API header files, the NuGet package structure, the props files and targets files to ensure the project will compile under Visual Studio + NuGet.

1> Create config file for NuGet -> https://github.com/Microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/runtest.bat#L25 

2> Restore packages  here -> https://github.com/Microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/runtest.bat#L55

3> MSBuild invocation here -> https://github.com/Microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/runtest.bat#L63

4> Run the sample here -> https://github.com/Microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/runtest.bat#L74